go/arrow/memory/internal/cgoalloc/allocator.cc

```diff
@@ -0,0 +1,75 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// +build ccalloc
+
+#include "allocator.h"
+#include "arrow/memory_pool.h"
+#include "helpers.h"
+
+struct mem_holder {
+    std::unique_ptr<arrow::MemoryPool> pool;
+    arrow::MemoryPool* current_pool;
```

Comment:
Why isn't this a `unique_ptr` as well?

Reply:
`arrow::default_memory_pool` returns a raw pointer, so if this is holding a reference to the *default* memory pool, `current_pool` can be set to the pointer without worry of deleting it. The `unique_ptr` is there for the case where we're not using the `default_memory_pool`, allowing the logic to not have to explicitly check which is in use when using the allocate/reallocate functions since the raw pointer should always be valid regardless of if it's using the default memory pool or not.

Reply:
I've added a comment explaining this, and also functions for referencing the default memory pool (which was in my original version of this but didn't make it into this cleaned up branch for whatever reason)

Reply:
But you're not using `default_memory_pool`, you're using `MemoryPool::CreateDefault`.

Reply:
I actually added a function which exposes the `default_memory_pool` and adds it as an optional pull in the package because it seemed like that made sense to do.

Reply:
Perhaps, but that doesn't change my concern. `CreateDefault` creates a _new_ memory pool instance, but you're not destroying it if logging is disabled.

Reply:
`holder->pool` is the `unique_ptr`, so when logging is disabled the call to `release_ref<mem_holder>(pool)` would destroy the `mem_holder` which would also destroy the memory pool instance created by `CreateDefault` when the `unique_ptr` destructor runs. Unless I'm missing something?

Reply:
Hmm, yes, you're right.
Still, this seems a bit convoluted. Why not:
```c++
struct mem_holder {
    std::unique_ptr<arrow::MemoryPool> wrapped_pool, current_pool;
};

ArrowMemoryPool arrow_create_memory_pool(bool enable_logging) {
    auto holder = std::make_shared<mem_holder>();
    if (enable_logging) {
        holder->wrapped_pool.reset(arrow::MemoryPool::CreateDefault());
        holder->current_pool.reset(new arrow::LoggingMemoryPool(holder->wrapped_pool.get()));
    } else {
        holder->current_pool.reset(arrow::MemoryPool::CreateDefault());
    }
    return create_ref(holder);
}

void arrow_release_pool(ArrowMemoryPool pool) {
    release_ref<mem_holder>(pool);
}
```

or even:
```c++
struct mem_holder {
    std::unique_ptr<arrow::MemoryPool> owned_pool;
    arrow::MemoryPool* pool;
};

ArrowMemoryPool arrow_create_memory_pool(bool enable_logging) {
    auto holder = std::make_shared<mem_holder>();
    if (enable_logging) {
        holder->owned_pool.reset(new arrow::LoggingMemoryPool(arrow::default_memory_pool()));
        holder->pool = holder->owned_pool.get();
    } else {
        holder->pool = arrow::default_memory_pool();
    }
    return create_ref(holder);
}

void arrow_release_pool(ArrowMemoryPool pool) {
    release_ref<mem_holder>(pool);
}
```

Reply:
I do like that as a better simplification, nice thanks! Can you think of any reason that it might be better to use `CreateDefault()`? Or does it make more sense to just use the `default_memory_pool` anyways?

Reply:
I can't think of a reason to use `CreateDefault`, unless for consistency you really need a `unique_ptr`. The `default_memory_pool` is destroyed at process exit.